src/coreclr/jit/lower.cpp

```diff
@@ -1994,9 +1994,107 @@ void Lowering::LowerArgsForCall(GenTreeCall* call)
         LowerArg(call, &arg, true);
     }
 
+    LowerSpecialCopyArgs(call);
+
     LegalizeArgPlacement(call);
 }
 
+void Lowering::LowerSpecialCopyArgs(GenTreeCall* call)
+{
+    // We only need to use the special copy helper on P/Invoke IL stubs
+    // for the unmanaged call.
+    if (comp->opts.jitFlags->IsSet(JitFlags::JIT_FLAG_IL_STUB) && comp->compMethodRequiresPInvokeFrame() &&
+        call->IsUnmanaged())
+    {
+        unsigned argIndex = 0;
+        for (CallArg& arg : call->gtArgs.Args())
+        {
+            if (!arg.IsUserArg())
+            {
+                continue;
+            }
+
+            if (argIndex >= comp->info.compArgsCount)
+            {
+                break;
+            }
+
+            // check if parameter at the same index as the IL argument is marked as requiring special copy, assuming
+            // that it is being passed 1:1 to the pinvoke
+            unsigned paramIndex = comp->compMap2ILvarNum(argIndex);
+            if ((paramIndex != (unsigned)ICorDebugInfo::UNKNOWN_ILNUM) && comp->argRequiresSpecialCopy(paramIndex) &&
+                arg.GetSignatureType() == TYP_STRUCT)
```

Comment:
Sorry, I think I gave you the wrong helper function to use here. I think this should rather be:

```suggestion
            unsigned paramLclNum = comp->compMapILargNum(argIndex);
            assert(paramLclNum < comp->info.compArgsCount);

            // check if parameter at the same index as the IL argument is marked as requiring special copy, assuming
            // that it is being passed 1:1 to the pinvoke
            if (comp->argRequiresSpecialCopy(paramLclNum) && (arg.GetSignatureType() == TYP_STRUCT))
```

Reply:
How is win-x64 handled in the new scheme? It looks like the JIT is not handling it. (I expect the `GT_PUTARG_STK` assert would be hit below otherwise.)

Reply:
win-x64 ends up passing pointers instead of structs. That's where the GetSignatureType came from (before that we hit the assert).

Reply:
Can you put the code to handle inserting the copy constructor calls under `TARGET_X86` then?